### PR TITLE
[IA-4364] Updating docker images to python 3.10

### PR DESCRIPTION
**Pull request by lucymcnatt**
__Jira ticket__: https://broadworkbench.atlassian.net/browse/IA-4364
__Jira ticket__: https://broadworkbench.atlassian.net/browse/IA-3994
__Jira ticket__: https://broadworkbench.atlassian.net/browse/IA-4271
__Jira ticket__: https://broadworkbench.atlassian.net/browse/IA-4304
__Jira ticket__: https://broadworkbench.atlassian.net/browse/IA-4356

 ## Dependencies
[Terra docker PR that generated the images](https://github.com/DataBiosphere/terra-docker/pull/434/files)

### What

- Updates the base image from` tf-gpu.2-7` to `tf-gpu.2-11.py310` to use python 3.10
- Removes references to gatk-otvf image
- Enables toc2 jupyter extension to allow file downloads in different formats
- cromshell 2.0 fix

### Why

- Python 3.7 is EOL (no more bug patches) at the end of June
- Intel decided to stop development of OVTF and will no longer be maintained
- Users want to save their notebooks as HTML or other formats

### What to test

Choose an image:
- create a jupyter notebook with their image and default settings
- run through the T101 notebook to make sure basic functionality works (can be in any workspace)
- https://app.terra.bio/#workspaces/help-terra/T101-Notebooks-Quickstart/analysis/launch/Analyze-data-from-table.ipynb
- copy the workspace of their featured notebook
    - create a Jupyter notebook with their image
    - run all cells for their respective notebook(s)
- try downloading their notebook as a non-IPYNB file

* R/Bioconductor
    * us.gcr.io/broad-dsp-gcr-public/terra-jupyter-bioconductor:2.2.1
    * https://app.terra.bio/#workspaces/help-gatk/Bioconductor/analysis/launch/SingleCellExperiment.ipynb
    * https://app.terra.bio/#workspaces/help-gatk/Bioconductor/analysis/launch/edgeR.ipynb
* Hail
    * us.gcr.io/broad-dsp-gcr-public/terra-jupyter-hail:1.1.1
    * https://app.terra.bio/#workspaces/help-gatk/Hail-Notebook-Tutorials/analysis/launch/01-genome-wide-association-study.ipynb 
    * https://app.terra.bio/#workspaces/featured-workspaces-hca/Intro-to-HCA-data-on-Terra/analysis/launch/07-matrixtable.ipynb
* Python
    * us.gcr.io/broad-dsp-gcr-public/terra-jupyter-python:1.1.1
    * https://app.terra.bio/#workspaces/featured-workspaces-hca/Intro-to-HCA-data-on-Terra/analysis/launch/Pegasus.ipynb 
* Default
    * us.gcr.io/broad-dsp-gcr-public/terra-jupyter-gatk:2.3.1
    * https://app.terra.bio/#workspaces/pathogen-genomic-surveillance/COVID-19/analyses
    * https://app.terra.bio/#workspaces/broad-firecloud-dsde/Hyperparameter%20Tuning%20with%20Optuna/analyses 

### Who tested and where

- [ ] This change is covered by automated tests 
  - _NB: Rerun automation tests on this PR by commenting `jenkins retest` or `jenkins multi-test`._
- [ ] I validated this change 
- [ ] Primary reviewer validated this change 
- [ ] I validated this change in the dev environment

### CODE REVIEW

File: jenkins/dataproc-custom-images/prepare-custom-leonardo-jupyter-dataproc-image.sh [hunk 1]
```diff
@@ -37,7 +37,7 @@ docker_image_var_names="welder_server terra_jupyter_python terra_jupyter_r terra
 
 # The version of python to install
 # Note: this should match the version of python in the terra-jupyter-hail image.
-python_version="3.7.10"
+python_version="3.10.12" // TODO LM (check on this)
```
Comment (LizBaldo) [inline jenkins/dataproc-custom-images/prepare-custom-leonardo-jupyter-dataproc-image.sh]: I think we have 3.10.9 in the docker image

File: http/src/main/resources/reference.conf [hunk 2]
```diff
@@ -640,7 +640,7 @@ image {
   welderGcrUri = "us.gcr.io/broad-dsp-gcr-public/welder-server"
   welderDockerHubUri = "broadinstitute/welder-server"
   welderHash = "6648f5c"
-  jupyterImage =  "us.gcr.io/broad-dsp-gcr-public/terra-jupyter-gatk:2.2.8"
+  jupyterImage =  "us.gcr.io/broad-dsp-gcr-public/terra-jupyter-gatk:2.3.0"
```
Comment (LizBaldo) [inline http/src/main/resources/reference.conf]: Ohhh good catch, i don't think that modifying this file is mentioned in the documentation :(

**Comment by LizBaldo** [comment]
jenkins retest

**Comment by LizBaldo** [comment]
jenkins retest

**Comment by LizBaldo** [comment]
jenkins retest

**Comment by LizBaldo** [comment]
jenkins retest

**Comment by LizBaldo** [comment]
jenkins retest

**Comment by LizBaldo** [comment]
jenkins retest

**Comment by LizBaldo** [comment]
jenkins retest

**Comment by LizBaldo** [comment]
jenkins retest

**Comment by LizBaldo** [comment]
jenkins retest

**Comment by LizBaldo** [comment]
jenkins retest

File: http/src/main/resources/reference.conf [hunk 3]
```diff
@@ -68,14 +68,16 @@ dataproc {
     region = "us-central1"
   }
 
-  customDataprocImage = "projects/broad-dsp-gcr-public/global/images/leo-dataproc-image-2-0-51-debian10-2023-06-09-19-40-21"
+  customDataprocImage = "projects/broad-dsp-gcr-public/global/images/leo-dataproc-image-2-1-11-debian11-2023-07-28-21-05-58"
```
Comment (LizBaldo) [inline http/src/main/resources/reference.conf:72]: @Qi77Qi this image does contain the latest AOU

File: http/src/main/resources/reference.conf [hunk 4]
```diff
@@ -68,14 +68,16 @@ dataproc {
...
+  legacyAouCustomDataprocImage = "projects/broad-dsp-gcr-public/global/images/leo-dataproc-image-2-0-51-debian10-2023-06-09-19-40-21"
```
Comment (LizBaldo) [inline http/src/main/resources/reference.conf:74]: @Qi77Qi this is the current cached dataproc image that contains the current AOU and therefore should not break anything for you

File: http/src/main/resources/reference.conf [hunk 5]
```diff
@@ -108,7 +110,7 @@ dataproc {
 }
 
 gce {
-  customGceImage = "projects/broad-dsp-gcr-public/global/images/leo-gce-image-2023-06-09-19-40-25"
+  customGceImage = "projects/broad-dsp-gcr-public/global/images/leo-gce-image-2023-08-01-20-11-52"
```
Comment (LizBaldo) [inline http/src/main/resources/reference.conf]: @Qi77Qi I regenerated a GCE image that is caching all of the new python 3.10 terra images, but is keeping the old AOU 3.7 image

File: http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/util/DataprocInterpreter.scala [hunk 6]
```diff
@@ -189,7 +189,14 @@ class DataprocInterpreter[F[_]: Parallel](
         initScripts = initScriptResources.map(resource => GcsPath(initBucketName, GcsObjectName(resource.asString)))
 
         // If we need to support 2 version of dataproc custom image, we'll update this
-        dataprocImage = config.dataprocConfig.customDataprocImage
+//        dataprocImage = config.dataprocConfig.customDataprocImage
+
+        // We need to maintain the old version of the dataproc image to uncouple the terra from the aou release
```
Comment (LizBaldo) [inline http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/util/DataprocInterpreter.scala:194]: @Qi77Qi this is the branching logic to ensure that we use the legacy dataproc image for AOU, we can remove once you have safely migrated

**Review by LizBaldo — COMMENTED** [review]
I need to sync my BEE and test that we can create runtimes with dataproc AND gce for the current aou image (2.1.22), but this should decouple the Terra image change from the AOU one

**Comment by LizBaldo** [comment]
jenkins retest

File: http/src/main/resources/reference.conf (hunk 3 above)
Comment (Qi77Qi) [reply http/src/main/resources/reference.conf:72]: do u mean does NOT?

File: http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/util/DataprocInterpreter.scala [hunk 7]
```diff
@@ -189,7 +189,14 @@ class DataprocInterpreter[F[_]: Parallel](
...
+        imageUrls = params.runtimeImages.map(_.imageUrl)
+        dataprocImage =
+          if (imageUrls.contains("us.gcr.io/broad-dsp-gcr-public/terra-jupyter-aou:2.1.22"))
```
Comment (Qi77Qi) [inline http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/util/DataprocInterpreter.scala]: this should be 2.1.15

File: http/src/main/resources/reference.conf (hunk 3 above)
Comment (LizBaldo) [reply http/src/main/resources/reference.conf:72]: No, `customDataprocImage` contains 2.2.1
`legacyAouCustomDataprocImage` is the once containing 2.1.22

But both also contain 2.1.15 so we should be good

Comment (LizBaldo) [reply http/src/main/resources/reference.conf:72]: I will add more comments around that in the code to clarify, it is getting a bit confusing I agree

File: http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/util/DataprocInterpreter.scala (hunk 7 above)
Comment (LizBaldo) [reply http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/util/DataprocInterpreter.scala]: Thanks! fixing now :)

**Review by Qi77Qi — APPROVED** [review]
thanks for working through this! 🎉

**Comment by LizBaldo** [comment]
jenkins retest

**Comment by LizBaldo** [comment]
jenkins retest

**Comment by LizBaldo** [comment]
jenkins retest

**Comment by LizBaldo** [comment]
jenkins retest

**Comment by LizBaldo** [comment]
The only failures are from the azure swatomation test suite, @lucymcnatt would you be comfortable merging to dev?

**Comment by LizBaldo** [comment]
See https://fc-jenkins.dsp-techops.broadinstitute.org/job/leonardo-fiab-test-runner/56619/consoleFull

**Pull request merged**